src/coreclr/vm/debugdebugger.cpp

```diff
@@ -1030,37 +1196,57 @@ void DebugStackTrace::Element::InitPass2()
-        //
-        // 1) The instruction that caused a hardware exception (div by zero, null ref, etc).
-        // 2) The instruction after the call to an internal runtime function (FCALL like IL_Throw, IL_Rethrow,
-        //    JIT_OverFlow, etc.) that caused a software exception.
-        // 3) The instruction after the call to a managed function (non-leaf node).
-        //
-        // #2 and #3 are the cases that need to adjust dwOffset because they point after the call instruction
-        // and may point to the next (incorrect) IL instruction/source line. If STEF_IP_ADJUSTED is set,
-        // IP/dwOffset has already be decremented so don't decrement it again.
-        //
-        // When the dwOffset needs to be adjusted it is a lot simpler to decrement instead of trying to figure out
-        // the beginning of the instruction. It is enough for GetILOffsetFromNative to return the IL offset of the
-        // instruction throwing the exception.
-        bool fAdjustOffset = (this->flags & STEF_IP_ADJUSTED) == 0 && this->dwOffset > 0;
-        bRes = g_pDebugInterface->GetILOffsetFromNative(
-            pFunc,
-            (LPCBYTE)this->ip,
-            fAdjustOffset ? this->dwOffset - STACKWALK_CONTROLPC_ADJUST_OFFSET : this->dwOffset,
-            &this->dwILOffset);
+        // Check the cache!
+        uint32_t dwILOffsetFromCache;
+        if (CheckNativeToILCache((void*)this->ip, fAdjustOffset, &dwILOffsetFromCache))
+        {
+            this->dwILOffset = dwILOffsetFromCache;
+            bRes = true;
+        }
+#ifdef DEBUGGING_SUPPORTED
+        else if (g_pDebugInterface)
+        {
+            // To get the source line number of the actual code that threw an exception, the dwOffset needs to be
+            // adjusted in certain cases when calculating the IL offset.
+            //
+            // The dwOffset of the stack frame points to either:
+            //
+            // 1) The instruction that caused a hardware exception (div by zero, null ref, etc).
+            // 2) The instruction after the call to an internal runtime function (FCALL like IL_Throw, IL_Rethrow,
+            //    JIT_OverFlow, etc.) that caused a software exception.
+            // 3) The instruction after the call to a managed function (non-leaf node).
+            //
+            // #2 and #3 are the cases that need to adjust dwOffset because they point after the call instruction
+            // and may point to the next (incorrect) IL instruction/source line. If STEF_IP_ADJUSTED is set,
+            // IP/dwOffset has already be decremented so don't decrement it again.
+            //
+            // When the dwOffset needs to be adjusted it is a lot simpler to decrement instead of trying to figure out
+            // the beginning of the instruction. It is enough for GetILOffsetFromNative to return the IL offset of the
+            // instruction throwing the exception.
+            bRes = g_pDebugInterface->GetILOffsetFromNative(
```

Comment:
I'm hoping in the future we could stop querying this from the debug interface and instead get it directly from the code manager, but I take it that is out-of-scope for the current change? That feels totally reasonable to me.